lua/nvim-treesitter/textobjects.lua

```diff
@@ -66,13 +66,150 @@ function M.select_textobject(query_string)
+                              { adjacent_textobject_range[3] + 1, adjacent_textobject_range[4] })
+    end
+  end
+end
+
+-- luacheck: push ignore 631
+M.goto_next_start = function(query_string) M.goto_adjacent(query_string, 'forward', 'start', not 'same_parent', 'overlap ok') end
+M.goto_next_end = function(query_string) M.goto_adjacent(query_string, 'forward', not 'start', not 'same_parent', 'overlap ok') end
+M.goto_previous_start = function(query_string) M.goto_adjacent(query_string, not 'forward', 'start', not 'same_parent', 'overlap ok') end
+M.goto_previous_end = function(query_string) M.goto_adjacent(query_string, not 'forward', not 'start', not 'same_parent', 'overlap ok') end
+-- luacheck: pop
+
+function M.next_textobject(node, query_string, same_parent, overlapping_range_ok, bufnr)
+  local node = node or ts_utils.get_node_at_cursor()
+  local bufnr = bufnr or api.nvim_get_current_buf()
+  if not node then return end
+
+  local _, _, node_end = node:end_()
+  local search_start, _
+  if overlapping_range_ok then
+    _, _, search_start = node:start()
+  else
+    _, _, search_start = node:end_()
+  end
+
+  local next_node = queries.find_best_match(bufnr,
+                                            query_string,
+                                            'textobjects',
+                                            function(match)
+                                              if match.node == node then return end
+                                              if not same_parent or node:parent() == match.node:parent() then
+                                                local _, _, start = match.node:start()
+                                                local _, _, end_ = match.node:end_()
+                                                return start > search_start and end_ >= node_end
+                                              end
+                                            end,
+                                            function(match)
+                                              local _, _, node_start = match.node:start()
+                                              return -node_start
+                                            end)
+
+  return next_node and next_node.node
+end
+
+function M.previous_textobject(node, query_string, same_parent, overlapping_range_ok, bufnr)
+  local node = node or ts_utils.get_node_at_cursor()
+  local bufnr = bufnr or api.nvim_get_current_buf()
+  if not node then return end
+
+  local _, _, node_start = node:start()
+  local search_end, _
+  if overlapping_range_ok then
+    _, _, search_end = node:end_()
+    search_end = search_end + 1
+  else
+    _, _, search_end = node:start()
+  end
+  local previous_node = queries.find_best_match(bufnr,
+                                                query_string,
+                                                'textobjects',
+                                                function(match)
```

Comment:
maybe you could define this function outside and make a closure to instantiate parameters that comes from the upper function ? it would be easier to read and understand ^^